src/bentoml_io/serde.py

```diff
@@ -0,0 +1,101 @@
+import abc
+import io
+import json
+import pickle
+import typing as t
+
+from pydantic import BaseModel
+from pydantic import RootModel
+
+T = t.TypeVar("T", bound=BaseModel)
+
+
+class Serde(abc.ABC):
+    media_type: str
+
+    @abc.abstractmethod
+    def serialize_model(self, model: BaseModel) -> bytes:
+        ...
+
+    @abc.abstractmethod
+    def deserialize_model(self, model_bytes: bytes, cls: type[T]) -> T:
+        ...
+
+    @abc.abstractmethod
+    def serialize(self, obj: t.Any) -> bytes:
+        ...
+
+    @abc.abstractmethod
+    def deserialize(self, obj_bytes: bytes) -> t.Any:
+        ...
+
+
+class JSONSerde(Serde):
+    media_type = "application/json"
+
+    def serialize_model(self, model: BaseModel) -> bytes:
+        return model.model_dump_json().encode("utf-8")
+
+    def deserialize_model(self, model_bytes: bytes, cls: type[T]) -> T:
+        return cls.model_validate_json(model_bytes)
+
+    def serialize(self, obj: t.Any) -> bytes:
+        return json.dumps(obj).encode("utf-8")
+
+    def deserialize(self, obj_bytes: bytes) -> t.Any:
+        return json.loads(obj_bytes)
+
+
+class PickleSerde(Serde):
+    media_type = "application/vnd.bentoml+pickle"
+
+    def serialize_model(self, model: BaseModel) -> bytes:
+        if isinstance(model, RootModel):
+            model_data: t.Any = model.root
+        else:
+            model_data = {k: getattr(model, k) for k in model.model_fields}
+        return pickle.dumps(model_data)
+
+    def deserialize_model(self, model_bytes: bytes, cls: type[T]) -> T:
+        obj = pickle.loads(model_bytes)
+        if not isinstance(obj, cls):
+            obj = cls.model_validate(obj)
+        return obj
+
+    def serialize(self, obj: t.Any) -> bytes:
+        return pickle.dumps(obj)
```

Comment:
https://peps.python.org/pep-0574/

worth to also implement something like `bentoml._internal.utils.pickle.pep574_dumps/pep574_loads` for `PickleSerde.serialize/deserialize`